{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in library(xgboost): there is no package called ‘xgboost’\n",
     "output_type": "error",
     "traceback": [
      "Error in library(xgboost): there is no package called ‘xgboost’\nTraceback:\n",
      "1. library(xgboost)",
      "2. stop(txt, domain = NA)"
     ]
    }
   ],
   "source": [
    "rm(list = ls())\n",
    "setwd(\"C:/Users/ashar/Dropbox/PPMI\")\n",
    "library(PPMI)\n",
    "####### Load the Prediction libraries ###################\n",
    "library(xgboost)\n",
    "library(randomForest)\n",
    "library(glmnet)\n",
    "library(mlr)\n",
    "library(caret)\n",
    "library(mlrHyperopt)\n",
    "library(mlrMBO)\n",
    "library(pROC)\n",
    "\n",
    "##### Load the data #########\n",
    "pd.complete <- read.csv(file = 'PPMI_complete.csv')\n",
    "##### Load the annotation\n",
    "\n",
    "#####Load the annotation ########\n",
    "anno <- read.csv(file = 'PatientCharacteristics_PPMI.csv')\n",
    "rownames(anno) <- anno$X\n",
    "\n",
    "##### Check whether the rownames are the same\n",
    "pd.complete[,1] == rownames(anno)\n",
    "\n",
    "############# INPUT SNP data ###########################################\n",
    "############# 3742 SNPs (based on some PASS filter) ###################\n",
    "snp.data <- read.csv(file = 'PPMI_combPD_genotype.csv')\n",
    "\n",
    "############# Get the names of the \n",
    "names.snp <- names(snp.data)\n",
    "\n",
    "names.snp2 <- c(0)\n",
    "names.snp2[1] <- \"X\"\n",
    "for( i in 2:length(names.snp)){\n",
    "  names.snp2[i] <- unlist(strsplit(as.character(names.snp[i]), split = 'PPMISI'))[[2]]\n",
    "}\n",
    "\n",
    "######## Check how many PD patients have corresponding SNP data available ##########\n",
    "###### 354 Patients have corresponding annotation ##################################\n",
    "sum((names.snp2 %in% rownames(anno))+0)\n",
    "\n",
    "################ Rename the Variables of the SNP data ################################\n",
    "names(snp.data) <- names.snp2\n",
    "\n",
    "############### Take only those patients which have corresponding annotation ##########\n",
    "snp.rel <- snp.data[names(snp.data) %in% pd.complete[,1] ]\n",
    "\n",
    "row.names(snp.rel) <- snp.data$X\n",
    "\n",
    "######################## Arrange the SNP data ##########\n",
    "group <- anno$PD.progression.HDDC[match(colnames(snp.rel), rownames(anno))]\n",
    "\n",
    "names(group) <- rownames(anno)[match(colnames(snp.rel), rownames(anno))]\n",
    "\n",
    "\n",
    "######################## Arrange the SNP data ##########\n",
    "rownames(anno) == pd.complete$X\n",
    "\n",
    "anno.snp <- anno[match(colnames(snp.rel), rownames(anno)),]\n",
    "\n",
    "colnames(snp.rel) == rownames(anno.snp)\n",
    "\n",
    "\n",
    "########## It's better to Parse out the Risk Allele Variant from the names #############\n",
    "all.snps <- c(0)\n",
    "for(i in 1:dim(snp.rel)[1]){\n",
    "  all.snps[i] <-  unlist(strsplit(rownames(snp.rel)[i], \"_\"))[1]\n",
    "}\n",
    "\n",
    "row.names(snp.rel) <- all.snps\n",
    "\n",
    "modality1 <- t(snp.rel)\n",
    "\n",
    "########## How can I visualize this using complex Heatmap library\n",
    "# library(ComplexHeatmap)\n",
    "# \n",
    "# split <- paste0(\"Cluster \", group)\n",
    "# df <- as.data.frame(as.character(group))\n",
    "# names(df) <- 'cluster'\n",
    "# cols <- gray.colors(3, start = 1.0, end = 0)\n",
    "# ha = HeatmapAnnotation(df = df, col = list(cluster = c(\"1\" = \"black\", \"2\" = \"red\", \"3\" = \"green\")), which = \"row\", width = unit(2, \"cm\"))\n",
    "# ht <- Heatmap(t(snp.rel),col = cols, split=split, cluster_columns = TRUE, show_heatmap_legend = T, show_row_names =FALSE, column_title = \"SNP mutation\", row_title = \"SNP\", name = 'SNP data', column_title_side = \"top\")\n",
    "\n",
    "# pdf('SNP_viz.pdf')\n",
    "# ht+ ha\n",
    "# dev.off()\n",
    "\n",
    "######### Load the SNP annotation ####################\n",
    "load(\"C:/Users/ashar/Dropbox/PPMI/SNP/RiskSNPs.rda\")\n",
    "\n",
    "genes.pd <- names(table(combPD$GENCODE_name))\n",
    "\n",
    "##### Load the 18 mechanism based SNPs ##########################\n",
    "##### Check if ALL 18 SNPs are contained in the SNP matrix ######\n",
    "subgraphPD.snps.rel <- list(NA)\n",
    "for( i in 1:length(subgraphPD.snps)){\n",
    "  snps <- subgraphPD.snps[[i]]\n",
    "  snps.real <- snps[snps %in% rownames(snp.rel) ]\n",
    "  subgraphPD.snps.rel[[i]] <- snps.real\n",
    "  names(subgraphPD.snps.rel)[i] <- names(subgraphPD.snps)[i]\n",
    "}\n",
    "\n",
    "##18 Mechanisms based SNPS ##############\n",
    "Y.list <- list(NA)\n",
    "for( i in 1:length(subgraphPD.snps)){\n",
    "  snps <- subgraphPD.snps[[i]]\n",
    "  Y.list[[i]] <- t(snp.rel[subgraphPD.snps.rel[[i]],])\n",
    "  names(Y.list)[i] <- names(subgraphPD.snps)[i]\n",
    "}\n",
    "\n",
    "###### Better the SNP matrix #############\n",
    "##### Check if SNP columns are NA ########\n",
    "Y.list.rel <- list(NA)\n",
    "for( i in 1:length(Y.list)){\n",
    "  indx <- c(0)\n",
    "  tmp <- Y.list[[i]]\n",
    "  indx <- as.numeric(which(is.na(tmp[1,])))\n",
    "  if(length(indx) == dim(tmp)[2]){\n",
    "    Y.list.rel[[i]] <- NA\n",
    "  } else if(length(indx)==0){\n",
    "    Y.list.rel[[i]] <- tmp\n",
    "  } else{\n",
    "    Y.list.rel[[i]] <- tmp[,-indx] \n",
    "  }\n",
    "  names(Y.list.rel)[i] <- names(subgraphPD.snps)[i]\n",
    "}\n",
    "\n",
    "####### Remove the list which is NA ##################################\n",
    "####### This means that there are 13 Mechanism without NA ############\n",
    "Y.list.final <- list(NA)\n",
    "countr = 1\n",
    "for ( j in 1:length(Y.list.rel)){\n",
    "  if(!is.na(Y.list.rel[[j]])){\n",
    "    Y.list.final[[countr]] <- Y.list.rel[[j]]\n",
    "    names(Y.list.final)[countr] <- names(Y.list.rel)[j]\n",
    "    countr = countr + 1\n",
    "  } \n",
    "}\n",
    "\n",
    "modality2 <- Y.list.final\n",
    "\n",
    "\n",
    "###### Check Names of List#######\n",
    "subgraphPD.rel <- subgraphPD[names(Y.list.final)]\n",
    "\n",
    "\n",
    "############################################################################################################\n",
    "############## Let's see if the Disease SNPs are the same as downloaded from GWAS catalogue from EBI ########\n",
    "#############################################################################################################\n",
    "# \n",
    "# emblSNPS <- read.csv(file = \"C:/Users/ashar/Dropbox/PPMI/SNP/SNPs_embl.csv\")\n",
    "# \n",
    "# ##### 185 disease associated SNPs ##############################\n",
    "# snps.embl <- emblSNPS$SNPS[emblSNPS$SNPS %in% rownames(snp.rel)]\n",
    "\n",
    "###### Get the disease related SNPS ####################################\n",
    "riskSNPS <- read.csv(file = \"C:/Users/ashar/Dropbox/PPMI/SNP/CuratedSNPs_DisGeNET_PD.csv\")\n",
    "snps.risk <- as.character(unique(riskSNPS$SNP[(riskSNPS$SNP %in% rownames(snp.rel))]))\n",
    "\n",
    "########## Modality 3 #################################################\n",
    "modality3 <- modality1[,colnames(modality1) %in% snps.risk]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
